File: ClientAPI/src/main/java/pt/upskill/clientapi/Controllers/CategoryController.java
package pt.upskill.clientapi.Controllers;

import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.media.Content;
import io.swagger.v3.oas.annotations.media.Schema;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.responses.ApiResponses;
import io.swagger.v3.oas.annotations.tags.Tag;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import pt.upskill.clientapi.CustomExceptions.CategoryNotFoundException;
import pt.upskill.clientapi.DTOs.CategoryDTO;
import pt.upskill.clientapi.DTOs.ResponseMessage;
import pt.upskill.clientapi.Interfaces.CategoryDAO;
import pt.upskill.clientapi.Mappers.CategoryMapper;
import pt.upskill.clientapi.Models.Category;

import java.util.List;

/**
 * Controller responsible for managing categories. This includes creating, retrieving,
 * and deleting categories, along with handling category-related operations.
 */
@Tag(name = "Category Management", description = "Operations related to categories")
@RestController
@RequestMapping("api/category")
public class CategoryController {

    private final CategoryDAO service;

    /**
     * Constructor to initialize CategoryController.
     *
     * @param service The CategoryDAO service used to interact with the category data.
     */
    public CategoryController(CategoryDAO service) {
        this.service = service;
    }

    /**
     * Endpoint to create a new category. The category includes a description.
     *
     * @param dto The CategoryDTO containing the description for the new category.
     * @return A ResponseEntity containing a message and HTTP status indicating success or failure.
     */
    @Operation(summary = "Create a new Category", description = "Creates a new category with a description.")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "201", description = "Category created successfully",
                    content = @Content(schema = @Schema(implementation = ResponseMessage.class))),
            @ApiResponse(responseCode = "400", description = "Invalid input",
                    content = @Content(schema = @Schema(implementation = ResponseMessage.class)))
    })
    @PostMapping
    public ResponseEntity<ResponseMessage> createCategory(@RequestBody CategoryDTO dto) {
        try {
            Category category = CategoryMapper.fromDTO(dto);
            Category newCategory = service.createCategory(category);
            return new ResponseEntity<>(new ResponseMessage(String.format("Category %s created successfully", newCategory.getDescription())), HttpStatus.CREATED);
        } catch (IllegalArgumentException e) {
            return new ResponseEntity<>(new ResponseMessage(e.getMessage()), HttpStatus.BAD_REQUEST);
        } catch (Exception e) {
            return new ResponseEntity<>(new ResponseMessage("Something went wrong"), HttpStatus.BAD_REQUEST);
        }
    }

    /**
     * Endpoint to retrieve all available categories.
     *
     * @return A ResponseEntity containing a list of all categories, or a message indicating no categories were found.
     */
    @Operation(summary = "Get all Categories", description = "Fetches a list of all available categories.")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", description = "List of categories retrieved successfully",
                    content = @Content),
            @ApiResponse(responseCode = "204", description = "No categories found",
                    content = @Content(schema = @Schema(implementation = ResponseMessage.class))),
            @ApiResponse(responseCode = "400", description = "Something went wrong",
                    content = @Content(schema = @Schema(implementation = ResponseMessage.class)))
    })
    @GetMapping
    public ResponseEntity<?> getAllCategories() {
        try {
            List<CategoryDTO> categories = service.getAllCategories().stream().map(CategoryMapper::toDTO).toList();
            return (categories.isEmpty()) ? new ResponseEntity<>(new ResponseMessage("No categories found"), HttpStatus.NO_CONTENT) : new ResponseEntity<>(categories, HttpStatus.OK);
        } catch (Exception e) {
            return new ResponseEntity<>(new ResponseMessage("Something went wrong"), HttpStatus.BAD_REQUEST);
        }
    }

    /**
     * Endpoint to retrieve a category by its description.
     *
     * @param description The description of the category to be fetched.
     * @return A ResponseEntity containing the category, or an error message if not found.
     */
    @Operation(summary = "Get a Category by Description", description = "Fetches a category by its description.")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", description = "Category retrieved successfully",
                    content = @Content(schema = @Schema(implementation = CategoryDTO.class))),
            @ApiResponse(responseCode = "404", description = "Category not found",
                    content = @Content(schema = @Schema(implementation = ResponseMessage.class))),
            @ApiResponse(responseCode = "400", description = "Something went wrong",
                    content = @Content(schema = @Schema(implementation = ResponseMessage.class)))
    })
    @GetMapping("{description}")
    public ResponseEntity<?> getCategory(@PathVariable String description) {
        try {
            Category category = service.getCategoryByDescription(description);
            CategoryDTO dto = CategoryMapper.toDTO(category);
            return new ResponseEntity<>(dto, HttpStatus.OK);
        } catch (CategoryNotFoundException e) {
            return new ResponseEntity<>(new ResponseMessage(e.getMessage()), HttpStatus.NOT_FOUND);
        } catch (Exception e) {
            return new ResponseEntity<>(new ResponseMessage("Something went wrong"), HttpStatus.BAD_REQUEST);
        }
    }

    /**
     * Endpoint to delete a category by its description.
     *
     * @param description The description of the category to be deleted.
     * @return A ResponseEntity indicating whether the deletion was successful or not.
     */
    @Operation(summary = "Delete a Category", description = "Deletes a category by its description.")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "204", description = "Category deleted successfully"),
            @ApiResponse(responseCode = "404", description = "Category not found",
                    content = @Content(schema = @Schema(implementation = ResponseMessage.class))),
            @ApiResponse(responseCode = "400", description = "Something went wrong",
                    content = @Content(schema = @Schema(implementation = ResponseMessage.class)))
    })
    @DeleteMapping("{description}")
    public ResponseEntity<ResponseMessage> deleteCategory(@PathVariable String description) {
        try {
            service.deleteCategoryByDescription(description);
            return ResponseEntity.noContent().build();
        } catch (CategoryNotFoundException e) {
            return new ResponseEntity<>(new ResponseMessage(e.getMessage()), HttpStatus.NOT_FOUND);
        } catch (Exception e) {
            return new ResponseEntity<>(new ResponseMessage("Something went wrong"), HttpStatus.BAD_REQUEST);
        }
    }
}

File: ForumAPI/ForumAPI/Controllers/CategoryController.cs
using ForumAPI.DTOs.CategoryDTOs;
using ForumAPI.Interfaces;

namespace ForumAPI.Controllers
{
    using Microsoft.AspNetCore.Mvc;
    using Swashbuckle.AspNetCore.Annotations;

    [Route("api/[controller]")]
    [ApiController]
    [SwaggerTag("Category Management")]
    public class CategoryController : ControllerBase
    {
        private readonly CategoryDAO service;

        /// <summary>
        /// Initializes a new instance of the <see cref="CategoryController"/> class.
        /// </summary>
        /// <param name="service">The category service.</param>
        public CategoryController(CategoryDAO service)
        {
            this.service = service;
        }

        /// <summary>
        /// Retrieves all categories.
        /// </summary>
        /// <remarks>
        /// This endpoint returns a list of all categories available in the system.
        /// If no categories exist, a 204 No Content status will be returned.
        /// </remarks>
        /// <returns>An <see cref="IActionResult"/> containing the list of categories or a No Content status.</returns>
        [HttpGet]
        [SwaggerOperation(
            Summary = "Retrieve all categories",
            Description = "This endpoint fetches all categories from the database. Returns a 204 No Content if no categories are found."
        )]
        [SwaggerResponse(200, "List of categories returned successfully.")]
        [SwaggerResponse(204, "No categories found.")]
        [SwaggerResponse(400, "Bad request.")]
        public async Task<IActionResult> GetAllCategories()
        {
            try
            {
                var categories = await service.GetAllCategories();
                return (categories.Any()) ? Ok(categories) : NoContent();
            }
            catch (ResponseStatusException ex)
            {
                return StatusCode((int)ex.StatusCode, ex.ResponseMessage);
            }
            catch (Exception ex)
            {
                return StatusCode(400, "Something went wrong");
            }
        }

        /// <summary>
        /// Retrieves a category by its description.
        /// </summary>
        /// <param name="description">The description of the category to retrieve.</param>
        /// <remarks>
        /// This endpoint allows the retrieval of a category based on its description.
        /// </remarks>
        /// <returns>An <see cref="IActionResult"/> containing the category details.</returns>
        [HttpGet("{description}")]
        [SwaggerOperation(
            Summary = "Retrieve category by description",
            Description = "This endpoint retrieves a category by its description."
        )]
        [SwaggerResponse(200, "Category found successfully.")]
        [SwaggerResponse(400, "Bad request.")]
        [SwaggerResponse(404, "Category not found.")]
        public async Task<IActionResult> GetCategoryByDescription(string description)
        {
            try
            {
                var category = await service.GetCategoryByDescription(description);
                return Ok(category);
            }
            catch (ResponseStatusException ex)
            {
                return StatusCode((int)ex.StatusCode, ex.ResponseMessage);
            }
            catch (Exception ex)
            {
                return StatusCode(400, "Something went wrong");
            }
        }

        /// <summary>
        /// Creates a new category.
        /// </summary>
        /// <param name="category">The category data to be created.</param>
        /// <remarks>
        /// This endpoint allows the creation of a new category.
        /// A successful creation returns a 201 Created status with the created category.
        /// </remarks>
        /// <returns>An <see cref="IActionResult"/> indicating the outcome of the category creation.</returns>
        [HttpPost]
        [SwaggerOperation(
            Summary = "Create a new category",
            Description = "This endpoint creates a new category in the database."
        )]
        [SwaggerResponse(201, "Category created successfully.")]
        [SwaggerResponse(400, "Bad request.")]
        public async Task<IActionResult> CreateCategory([FromBody] CategoryDTO category)
        {
            try
            {
                var message = await service.CreateCategory(category);
                return CreatedAtAction(nameof(CreateCategory), message);
            }
            catch (ResponseStatusException ex)
            {
                return StatusCode((int)ex.StatusCode, ex.ResponseMessage);
            }
            catch (Exception ex)
            {
                return StatusCode(400, "Something went wrong");
            }
        }

        /// <summary>
        /// Deletes a category by its description.
        /// </summary>
        /// <param name="description">The description of the category to delete.</param>
        /// <remarks>
        /// This endpoint deletes a category based on its description. A successful deletion returns a 204 No Content status.
        /// </remarks>
        /// <returns>An <see cref="IActionResult"/> indicating the outcome of the category deletion.</returns>
        [HttpDelete("{description}")]
        [SwaggerOperation(
            Summary = "Delete category by description",
            Description = "This endpoint deletes a category from the database based on its description."
        )]
        [SwaggerResponse(204, "Category deleted successfully.")]
        [SwaggerResponse(400, "Bad request.")]
        [SwaggerResponse(404, "Category not found.")]
        public async Task<IActionResult> DeleteCategory(string description)
        {
            try
            {
                await service.DeleteCategory(description);
                return NoContent();
            }
            catch (ResponseStatusException ex)
            {
                return StatusCode((int)ex.StatusCode, ex.ResponseMessage);
            }
            catch (Exception ex)
            {
                return StatusCode(400, "Something went wrong");
            }
        }
    }
}
